### PR TITLE
Rework storage accessibility checking

### LINKED ISSUE #33: Check Minio connection more intelligently
Currently, we check if Minio is accessible when shepherd is starting, i.e. at a time when it 1) doesn't really matter that much 2) is likely that it is in fact not accessible. This leads to some horrible hacks such as putting `sleep` in our Dockerfiles. Also, if Minio breaks down later, we might find out only when users start complaining.

I propose we think of a better way of checking Minio connectivity and reporting failures.

### PR DESCRIPTION
**Pull request by Teyras**
closes #33

### CODE REVIEW

File: shepherd/docker/registry.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import requests
+from requests.auth import HTTPBasicAuth
+
+from shepherd.shepherd.config import RegistryConfig
+
+
+def list_images_in_registry(config: RegistryConfig):
```
Comment (blazekadam) [inline shepherd/docker/registry.py]: Missing type annotations and docstring.

File: shepherd/shepherd/shepherd.py [hunk 2]
```diff
@@ -116,7 +121,25 @@ def enqueue_job(self, job_id: str, job_meta: ModelModel, sheep_id: Optional[str]
         self[sheep_id].jobs_meta[job_id] = job_meta
         self[sheep_id].jobs_queue.put(job_id)
 
-    def health_check(self, sheep_id: str) -> None:
+    def shepherd_health_check(self) -> None:
```
Comment (blazekadam) [inline shepherd/shepherd/shepherd.py]: Missing docstring.

File: shepherd/shepherd/shepherd.py [hunk 3]
```diff
@@ -116,7 +121,25 @@ def enqueue_job(self, job_id: str, job_meta: ModelModel, sheep_id: Optional[str]
...
+        while True:
+            gevent.sleep(1)
+
+            if not self.storage.is_accessible() and not self.storage_inaccessible_reported:
```
Comment (blazekadam) [inline shepherd/shepherd/shepherd.py]: The flag property should be protected, should it not?

File: shepherd/shepherd/shepherd.py [hunk 4]
```diff
@@ -116,7 +121,25 @@ def enqueue_job(self, job_id: str, job_meta: ModelModel, sheep_id: Optional[str]
...
+                self.storage_inaccessible_reported = True
+            else:
+                self.storage_inaccessible_reported = False
+
+            if self.registry_config is not None:
+                try:
+                    list_images_in_registry(self.registry_config)
+                    self.registry_inaccessible_reported = False
+                except:
+                    logging.warning("The Docker registry is not accessible")
```
Comment (blazekadam) [inline shepherd/shepherd/shepherd.py]: This is rather `error` level thing, is it not?

File: shepherd/manage.py [hunk 5]
```diff
@@ -50,14 +48,6 @@ def run(host, port, config_file) -> None:
     api_server = WSGIServer((host, port), app, log=logging.getLogger(''))
     api_handler = gevent.spawn(api_server.serve_forever)
 
-    # everything should be ready, lets check if minio works
```
Comment (blazekadam) [inline shepherd/manage.py]: Is there a solid reasoning behind removing this check? It was quite convenient as you would get immediate error if you forget to start the storage.

File: shepherd/storage/minio_storage.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
+from io import BytesIO
+
+from minio import Minio
+
+from shepherd.constants import DONE_FILE, ERROR_FILE
+from shepherd.storage.storage import Storage
+from shepherd.utils import minio_object_exists, pull_minio_bucket, push_minio_bucket
+
+
+class MinioStorage(Storage):
```
Comment (blazekadam) [inline shepherd/storage/minio_storage.py]: Missing docstrings in the whole file.

File: shepherd/storage/minio_storage.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        except BaseException:
+            return False
+
+    def pull_job_data(self, job_id: str, target_directory: str):
+        pull_minio_bucket(self._minio, job_id, target_directory)
+
+    def push_job_data(self, job_id: str, source_directory: str):
+        push_minio_bucket(self._minio, job_id, source_directory)
+
+    def job_failed(self, job_id: str, message: str):
```
Comment (blazekadam) [inline shepherd/storage/minio_storage.py]: A better name would be imo `report_job_failed`

File: shepherd/storage/minio_storage.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        error = message.encode()
+        self._minio.put_object(job_id, ERROR_FILE, BytesIO(error), len(error))
+
+    def job_done(self, job_id: str):
```
Comment (blazekadam) [inline shepherd/storage/minio_storage.py]: Similarly `report_job_done`?

File: shepherd/storage/storage.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+
+class Storage(metaclass=abc.ABCMeta):
+    """
+    An interface for services that provide access to job data in a remote storage
+    """
+
+    @abc.abstractmethod
+    def is_accessible(self) -> bool:
+        """
```
Comment (blazekadam) [inline shepherd/storage/storage.py]: One-liner docstrings may be contracted to """...""".

File: shepherd/storage/storage.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        Check if the remote storage can be accessed
+        """
+
+    @abc.abstractmethod
+    def pull_job_data(self, job_id: str, target_directory: str):
+        """
+        Download job files from the remote storage to a local directory
```
Comment (blazekadam) [inline shepherd/storage/storage.py]: Missing newline after the description, missing dot at the end of the sentence... (multiple times)

File: shepherd/storage/storage.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    """
+
+    @abc.abstractmethod
+    def is_accessible(self) -> bool:
+        """
+        Check if the remote storage can be accessed
+        """
+
+    @abc.abstractmethod
+    def pull_job_data(self, job_id: str, target_directory: str):
```
Comment (blazekadam) [inline shepherd/storage/storage.py]: Missing `-> None` (multiple times)

File: shepherd/storage/storage.py (hunk 10 above)
Comment (blazekadam) [reply shepherd/storage/storage.py]: Missing args description

File: shepherd/storage/minio_storage.py [hunk 12]
```diff
@@ -0,0 +1,38 @@
+from io import BytesIO
```
Comment (blazekadam) [inline shepherd/storage/minio_storage.py]: Pls expose `MinioStorage` and `Storage` in `init.py`'s `__all__`.

File: shepherd/shepherd/shepherd.py (hunk 4 above)
Comment (Teyras) [reply shepherd/shepherd/shepherd.py]: It might happen when we're launching a bunch of containers and Minio hasn't started yet, which is an error that will probably go away soon. This led me to a conclusion we don't need to be extra loud about this.

File: shepherd/manage.py (hunk 5 above)
Comment (Teyras) [reply shepherd/manage.py]: It's not convenient at all. In a docker-compose deployment, you need to start the shepherd with a delay, which is a) slow (Minio usually won't need the whole second) b) unreliable (when docker-compose is e.g. pulling a new minio image, one second might not be enough) c) ugly. If you forget to start Minio now, you'll get a stern warning without a crash - that's not much of a difference for a developer who launches the shepherd in a terminal.

In short, it's OK for Minio to not be accessible when shepherd starts, but it's absolutely not OK when it crashes after some time (think disk space running out). Until now, we checked exactly the opposite conditions.

File: shepherd/storage/minio_storage.py (hunk 7 above)
Comment (Teyras) [reply shepherd/storage/minio_storage.py]: Sure, what the hell.

File: shepherd/docker/registry.py [hunk 13]
```diff
@@ -0,0 +1,15 @@
+import requests
+from requests.auth import HTTPBasicAuth
+
+from ..config import RegistryConfig
+
+
+def list_images_in_registry(config: RegistryConfig) -> None:
```
Comment (blazekadam) [inline shepherd/docker/registry.py]: Inconsistent type annotation and missing return docstring.

File: shepherd/docker/utils.py [hunk 14]
```diff
@@ -2,7 +2,7 @@
 import subprocess
 from typing import List
 
-from .errors import DockerError
+from shepherd.errors.docker import DockerError
```
Comment (blazekadam) [inline shepherd/docker/utils.py]: maybe we want to use relative import here?

File: shepherd/manage.py (hunk 5 above)
Comment (blazekadam) [reply shepherd/manage.py]: Fair points.

File: shepherd/sheep/__init__.py [hunk 15]
```diff
@@ -1,6 +1,6 @@
 from .base_sheep import BaseSheep
 from .docker_sheep import DockerSheep
 from .bare_sheep import BareSheep
-from .errors import SheepError, SheepConfigurationError
+from ..errors.sheep import SheepError, SheepConfigurationError
 
 __all__ = ['BaseSheep',  'DockerSheep', 'BareSheep', 'SheepError', 'SheepConfigurationError']
```
Comment (blazekadam) [inline shepherd/sheep/__init__.py]: In `comm` you removed the errors from `__all__` while here you keep them. I would opt for the first option, anyways it should be consistent.

File: shepherd/sheep/docker_sheep.py [hunk 16]
```diff
@@ -4,9 +4,10 @@
 from schematics.types import BooleanType
 
 from .base_sheep import BaseSheep
-from .errors import SheepConfigurationError
-from ..docker import DockerContainer, DockerImage, DockerError
-from ..shepherd.config import RegistryConfig
+from shepherd.errors.sheep import SheepConfigurationError
```
Comment (blazekadam) [inline shepherd/sheep/docker_sheep.py]: any reason for the absolute import usage?

File: shepherd/shepherd/shepherd.py [hunk 17]
```diff
@@ -1,23 +1,23 @@
...
+from ..constants import OUTPUT_DIR
+from ..docker import list_images_in_registry
+from ..storage.minio_storage import MinioStorage
+from ..config import RegistryConfig
 from ..sheep import *
 from ..api.models import SheepModel
 from ..api.models import ModelModel
-from shepherd.sheep.errors import SheepConfigurationError
-from ..api.errors import UnknownSheepError, UnknownJobError
+from shepherd.errors.sheep import SheepConfigurationError
```
Comment (blazekadam) [inline shepherd/shepherd/shepherd.py]: ditto

File: shepherd/shepherd/shepherd.py [hunk 18]
```diff
@@ -64,9 +65,13 @@ def __init__(self,
             self.sheep[sheep_id] = sheep
             self.poller.register(socket, zmq.POLLIN)
             gevent.spawn(partial(self.dequeue_and_feed_jobs, sheep_id))
-            gevent.spawn(partial(self.health_check, sheep_id))
+            gevent.spawn(partial(self.sheep_health_check, sheep_id))
```
Comment (blazekadam) [inline shepherd/shepherd/shepherd.py]: Should not we store the greenlets and end them gracefully if need be? (e.g.: as we do on line 71)

File: shepherd/shepherd/shepherd.py [hunk 19]
```diff
@@ -116,7 +121,29 @@ def enqueue_job(self, job_id: str, job_meta: ModelModel, sheep_id: Optional[str]
...
+
+            if not self.storage.is_accessible() and not self._storage_inaccessible_reported:
+                logging.warning("The remote storage is not accessible")
+                self._storage_inaccessible_reported = True
+            else:
+                self._storage_inaccessible_reported = False
+
+            if self.registry_config is not None:
+                try:
+                    list_images_in_registry(self.registry_config)
```
Comment (blazekadam) [inline shepherd/shepherd/shepherd.py]:
```suggestion
                    _ = list_images_in_registry(self.registry_config)
```

File: shepherd/shepherd/shepherd.py (hunk 4 above)
Comment (blazekadam) [reply shepherd/shepherd/shepherd.py]: On the other hand, it is quite severe if it occurs later on... I would opt for the `error` level...

File: shepherd/storage/minio_storage.py [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+        try:
+            self._minio.list_buckets()
+            return True
+        except BaseException:
+            return False
+
+    is_accessible.__doc__ = Storage.is_accessible.__doc__
+
+    def pull_job_data(self, job_id: str, target_directory: str) -> None:
+        pull_minio_bucket(self._minio, job_id, target_directory)
```
Comment (blazekadam) [inline shepherd/storage/minio_storage.py]: We should wrap the `MinioError`s with our `StorageError`s (in all the `MinioStorage` functions).

File: shepherd/storage/minio_storage.py [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+        self._minio = minio
+
+    def is_accessible(self) -> bool:
+        try:
+            self._minio.list_buckets()
+            return True
+        except BaseException:
+            return False
+
+    is_accessible.__doc__ = Storage.is_accessible.__doc__
```
Comment (blazekadam) [inline shepherd/storage/minio_storage.py]: image

File: shepherd/utils/storage.py [hunk 22]
```diff
@@ -5,13 +5,15 @@
```
Comment (blazekadam) [inline shepherd/utils/storage.py]: Perhaps we can merge this file with the `MiniStorage` class?

File: shepherd/shepherd/shepherd.py (hunk 19 above)
Comment (Teyras) [reply shepherd/shepherd/shepherd.py]: Overruled by @petrbel and @FloopCZ

File: shepherd/utils/storage.py (hunk 22 above)
Comment (Teyras) [reply shepherd/utils/storage.py]: It certainly is a possibility. I didn't really want to introduce any more changes.

File: shepherd/storage/minio_storage.py (hunk 20 above)
Comment (Teyras) [reply shepherd/storage/minio_storage.py]: I just found out they are actually wrapped by `pull_minio_bucket` and friends.

File: shepherd/utils/storage.py (hunk 22 above)
Comment (Teyras) [reply shepherd/utils/storage.py]: On second thoughts, I'll go knees deep and refactor the crap out of it.

File: shepherd/shepherd/shepherd.py (hunk 19 above)
Comment (blazekadam) [reply shepherd/shepherd/shepherd.py]: better explicit than implicit?

File: shepherd/shepherd/shepherd.py [hunk 23]
```diff
@@ -104,6 +108,10 @@ def slaughter_sheep(self, sheep_id: str) -> None:
         :return:
         """
         logging.info('Slaughtering sheep `%s`', sheep_id)
+
+        for greenlet in self._sheep_greenlets[sheep_id]:
```
Comment (blazekadam) [inline shepherd/shepherd/shepherd.py]: Not sure if you want to kill the greenlets here. If it is the case, then you gonna need to start them every time you kill and (re)start the respective sheep.

File: shepherd/storage/minio_storage.py [hunk 24]
```diff
@@ -1,10 +1,34 @@
...
+from ..errors.api import StorageError, StorageInaccessibleError
+from ..constants import DONE_FILE, ERROR_FILE, INPUT_DIR, OUTPUT_DIR
+from .storage import Storage
+from ..utils import minio_object_exists
 
-from shepherd.constants import DONE_FILE, ERROR_FILE
-from shepherd.storage.storage import Storage
-from shepherd.utils import minio_object_exists, pull_minio_bucket, push_minio_bucket
+
+def inherit_doc(parent):
```
Comment (blazekadam) [inline shepherd/storage/minio_storage.py]: I am sorry for not suggesting it earlier but I would really prefer a simple `Implementing [sphinx link]` docsting over this...

File: shepherd/storage/minio_storage.py [hunk 25]
```diff
@@ -15,38 +39,78 @@ class MinioStorage(Storage):
...
+    @inherit_doc(Storage)
     def pull_job_data(self, job_id: str, target_directory: str) -> None:
-        pull_minio_bucket(self._minio, job_id, target_directory)
-
-    pull_job_data.__doc__ = Storage.pull_job_data.__doc__
+        logging.debug('Pulling minio bucket `%s` to dir `%s`', job_id, target_directory)
 
+        try:
+            pulled_count = 0
+            for object in self._minio.list_objects_v2(job_id, recursive=True):
```
Comment (blazekadam) [inline shepherd/storage/minio_storage.py]: Thank you for the refactor!!!

File: shepherd/shepherd/shepherd.py (hunk 23 above)
Comment (Teyras) [reply shepherd/shepherd/shepherd.py]: You're right, thanks.

File: shepherd/storage/minio_storage.py [hunk 26]
```diff
@@ -0,0 +1,120 @@
...
+
+class MinioStorage(Storage):
+    """
+    A remote storage adapter that uses the S3 client from Minio SDK.
+    """
+
+    def __init__(self, minio: Minio):
+        """
+        Initialize the storage with a Minio handle.
+        :param minio: Minio storage handle
```
Comment (blazekadam) [inline shepherd/storage/minio_storage.py]:
```suggestion

        :param minio: Minio storage handle
```
